lib/libconfigcpp.c++

```diff
@@ -126,182 +173,214 @@ static int __toTypeCode(Setting::Type type)
 
 // ---------------------------------------------------------------------------
 
-static void __constructPath(const Setting &setting,
-                            std::stringstream &path)
+static void __writeSettingPath(const Setting &setting, std::ostream &o)
 {
   // head recursion to print path from root to target
-
   if(! setting.isRoot())
   {
-    __constructPath(setting.getParent(), path);
-    if(path.tellp() > 0)
-      path << '.';
+    const Setting &parent_setting = setting.getParent();
+    __writeSettingPath(parent_setting, o);
+    if (! parent_setting.isRoot())
+      o << '.';
 
     const char *name = setting.getName();
+
     if(name)
-      path << name;
+        o << name;
     else
-      path << '[' << setting.getIndex() << ']';
+        o << '[' << setting.getIndex() << ']';
   }
 }
 
 // ---------------------------------------------------------------------------
 
-SettingException::SettingException(const Setting &setting)
+static std::string __constructSettingPath(const Setting &setting)
 {
-  std::stringstream sstr;
-  __constructPath(setting, sstr);
+  std::stringstream ss;
+  __writeSettingPath(setting, ss);
+  return ss.str();
+}
+
+// ---------------------------------------------------------------------------
 
-  _path = ::strdup(sstr.str().c_str());
+static std::string __constructSettingPath(const Setting &setting, int idx)
+{
+  std::stringstream ss;
+  __writeSettingPath(setting, ss);
+  ss << ".[" << idx << ']';
+  return ss.str();
 }
 
 // ---------------------------------------------------------------------------
 
-SettingException::SettingException(const Setting &setting, int idx)
+static std::string __constructSettingPath(const Setting &setting, const char *name)
 {
-  std::stringstream sstr;
-  __constructPath(setting, sstr);
-  sstr << ".[" << idx << "]";
+  std::stringstream ss;
+  __writeSettingPath(setting, ss);
+  ss << ".[" << name << ']';
```

Comment:
Don't use brackets here, as the path is 'parent.name', not 'parent.[name]'

Reply:
Is this notation for indexing?

Reply:
Brackets are used for indexing, like 'parent.[1]' but not for member access by name, like 'parent.child'